### PR TITLE
feat(WASM): Support setting the screen orientation

### LINKED ISSUE #3177: Add screen orientation support on WASM
## What would you like to be added:

Support for screen orientation via [Screen.orientation API](https://developer.mozilla.org/en-US/docs/Web/API/Screen/orientation).

Can also include screen orientation locking/unlocking (`AutoRotationPreferences` in UWP)

## Why is this needed:

Feature parity

## For which Platform:

- [ ] iOS
- [ ] Android
- [x] WebAssembly
- [ ] WebAssembly renderers for Xamarin.Forms
- [ ] macOS
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

### PR DESCRIPTION
**Pull request by trungnt2910**
Supports setting the screen orientation through the `DisplayInformation.AutoRotationPreferences` on supported browsers.

Internally the implementation calls the JavaScript `screen.orientation.lock` API.

GitHub Issue (If applicable): closes #3177

## PR Type

What kind of change does this PR introduce?

- Feature

## What is the current behavior?

Uno WASM applications ignore the `DisplayInformation.AutoRotationPreferences` property.

## What is the new behavior?

Uno WASM applications attempts to use the `screen.orientation.lock` JavaScript API to set the screen orientation on supported devices.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UWP/Graphics/Display/DisplayInformation.wasm.cs [hunk 1]
```diff
@@ -152,6 +161,24 @@ private static DisplayOrientations ParseJsOrientation(string jsOrientation)
...
+				DisplayOrientations.Portrait => "portrait-primary",
+				DisplayOrientations.PortraitFlipped => "portrait-secondary",
+				_ => "any"
+			};
+		}
+
+		private static Task SetOrientationAsync(DisplayOrientations orientations, CancellationToken ct)
+		{
+			var jsOrientation = WebAssemblyRuntime.EscapeJs(ToJsOrientation(orientations));
+			return WebAssemblyRuntime.InvokeAsync($"{JsType}.setOrientationAsync(\"{jsOrientation}\")", ct);
```
Comment (Youssef1313) [inline src/Uno.UWP/Graphics/Display/DisplayInformation.wasm.cs:180]: Not sure if it's worth doing a check to call this only if `jsOrientation` is different than the existing old value

Comment (trungnt2910) [reply src/Uno.UWP/Graphics/Display/DisplayInformation.wasm.cs:180]: To get the existing value we would need a Javascript interop call anyway.

File: src/Uno.UWP/Graphics/Display/DisplayInformation.wasm.cs [hunk 2]
```diff
@@ -152,6 +161,24 @@ private static DisplayOrientations ParseJsOrientation(string jsOrientation)
...
+		private static string ToJsOrientation(DisplayOrientations orientations)
+		{
+			return orientations switch
+			{
+				DisplayOrientations.Landscape => "landscape-primary",
+				DisplayOrientations.LandscapeFlipped => "landscape-secondary",
+				DisplayOrientations.Portrait => "portrait-primary",
+				DisplayOrientations.PortraitFlipped => "portrait-secondary",
+				_ => "any"
+			};
```
Comment (Youssef1313) [inline src/Uno.UWP/Graphics/Display/DisplayInformation.wasm.cs:174]: `DisplayOrientations` is a flags enum. So, multiple values can be specified.

I think we should well understand the meaning of them, and use the corresponding JS orientation.

File: src/Uno.UWP/Graphics/Display/DisplayInformation.wasm.cs [hunk 3]
```diff
@@ -152,6 +161,24 @@ private static DisplayOrientations ParseJsOrientation(string jsOrientation)
...
+
+		private static string ToJsOrientation(DisplayOrientations orientations)
+		{
+			return orientations switch
+			{
+				DisplayOrientations.Landscape => "landscape-primary",
+				DisplayOrientations.LandscapeFlipped => "landscape-secondary",
+				DisplayOrientations.Portrait => "portrait-primary",
+				DisplayOrientations.PortraitFlipped => "portrait-secondary",
+				_ => "any"
```
Comment (Youssef1313) [inline src/Uno.UWP/Graphics/Display/DisplayInformation.wasm.cs:173]: Did you find "any" in some documentation or just arbitrary string?

Comment (trungnt2910) [reply src/Uno.UWP/Graphics/Display/DisplayInformation.wasm.cs:173]: It's documented [here](https://developer.mozilla.org/en-US/docs/Web/API/ScreenOrientation/lock).

File: src/Uno.UWP/Graphics/Display/DisplayInformation.wasm.cs (hunk 2 above)
Comment (trungnt2910) [reply src/Uno.UWP/Graphics/Display/DisplayInformation.wasm.cs:174]: It seems like if multiple values are specified, the system gets to choose which orientation.

So `Portrait | PortraitFlipped` can be mapped to the JS `portrait`, while `Landscape | LanscapeFlipped` can be mapped to the JS `landscape`.

However, for other values, I don't think there's a suitable mapping. A solution might be passing a list to the TypeScript function, and then return when setting to one of the values succeed?

Comment (Youssef1313) [reply src/Uno.UWP/Graphics/Display/DisplayInformation.wasm.cs:174]: @MartinZikmund Could you advise here?

Comment (trungnt2910) [reply src/Uno.UWP/Graphics/Display/DisplayInformation.wasm.cs]: I've removed the mentioned code. Instead, I'm passing the value as-is to the JavaScript code (passing an int is much cheaper than passing an array of strings, which would require JSON serialization).

The JavaScript code would then decode the enum to a list of `OrientationLockType`, and attempt to lock the orientation at one of the types specified.

Comment (trungnt2910) [reply src/Uno.UWP/Graphics/Display/DisplayInformation.wasm.cs]: This code is outdated, so I'll mark this as resolved.

**Review by MartinZikmund — APPROVED** [review]
Very nice! It is interesting that browsers require full-screen 😀 . Approving!

**Pull request merged**